src/main/kotlin/de/smartsquare/starter/mqtt/MqttHandler.kt

```diff
@@ -18,57 +23,76 @@ class MqttHandler(
+     */
+    private fun getSubscriber(topic: MqttTopic): MqttSubscriber = subscriberCache.getOrPut(topic) {
+        val subscriber = collector.subscribers.find { it.topic.matches(topic) }
             ?: error("No subscriber found for topic $topic")
+        val kFunction = subscriber.method.kotlinFunction
+        val parameterTypes = kFunction?.valueParameters?.map { it.type.jvmErasure.java }
+            ?: subscriber.method.parameterTypes.toList()
+        val delegate = if (kFunction?.isSuspend == true) {
+            AnnotatedMethodDelegate { args -> runBlocking { kFunction.callSuspend(subscriber.bean, *args) } }
+        } else {
+            val handle = lookup.unreflect(subscriber.method)
+            AnnotatedMethodDelegate { args -> handle.invokeWithArguments(subscriber.bean, *args) }
+        }
 
-        try {
-            val parameters = subscriber.method.parameterTypes
-                .map { adapter.adapt(message, it) }
-                .toTypedArray()
+        MqttSubscriber(delegate, parameterTypes)
+    }
 
-            subscriber.method.invoke(subscriber.bean, *parameters)
-        } catch (e: InvocationTargetException) {
-            messageErrorHandler.handle(
-                MqttMessageException(
-                    topic,
-                    message.payload,
-                    "Error while handling mqtt message on topic [$topic]",
-                    e,
-                ),
-            )
+    /**
+     * Adapts the payload of the [message] to the [parameterTypes] of the subscriber.
+     * If an error occurs, the error is logged and the [MqttMessageErrorHandler] is called.
+     */
+    private fun adaptParameters(
+        parameterTypes: List<Class<*>>,
+        message: MqttPublishContainer,
+        topic: MqttTopic,
+    ): Result<Array<Any>> {
+        val (e, errorMessage) = try {
+            return Result.success(parameterTypes.map { adapter.adapt(message, it) }.toTypedArray())
         } catch (e: JsonMappingException) {
-            messageErrorHandler.handle(
-                MqttMessageException(
-                    topic,
-                    message.payload,
-                    "Error while handling mqtt message on topic [$topic]: Failed to map payload to target class",
-                    e,
-                ),
-            )
+            e to "Error while handling mqtt message on topic [$topic]: Failed to map payload to target class"
         } catch (e: JacksonException) {
-            messageErrorHandler.handle(
-                MqttMessageException(
-                    topic,
-                    message.payload,
-                    "Error while handling mqtt message on topic [$topic]: Failed to parse payload",
-                    e,
-                ),
-            )
+            e to "Error while handling mqtt message on topic [$topic]: Failed to parse payload"
         }
+
+        return Result.failure(MqttMessageException(topic, message.payload, errorMessage, e))
     }
```

Comment:
I think we should inline this back. It takes the `message` and `topic` parameters only to use them in the error handling and otherwise is only a single line of actual work.

I would also like to avoid mixing thrown exceptions and the `Result` class so that we only have one - thrown exceptions since that is what we use everywhere else.

Reply:
Mind, that I do catch for Exception, as MethodHandle, as well as kotlin reflect don't wrap any potential Exceptions thrown in a InvocationTargetException